pkg/device_plugin/generic_device_plugin.go

```diff
@@ -349,24 +380,105 @@ func (dpi *GenericDevicePlugin) PreStartContainer(ctx context.Context, in *plugi
 	return res, nil
 }
 
-// GetPreferredAllocation is for compatible with new DevicePluginServer API for DevicePlugin service. It has not been implemented in kubevrit-gpu-device-plugin
+// GetPreferredAllocation returns a preferred set of devices to allocate
+// from a list of available ones. This helps the Topology Manager make
+// topology-aware allocation decisions based on NUMA affinity.
 func (dpi *GenericDevicePlugin) GetPreferredAllocation(ctx context.Context, in *pluginapi.PreferredAllocationRequest) (*pluginapi.PreferredAllocationResponse, error) {
-	// TODO
-	// returns a preferred set of devices to allocate
-	// from a list of available ones. The resulting preferred allocation is not
-	// guaranteed to be the allocation ultimately performed by the
-	// devicemanager. It is only designed to help the devicemanager make a more
-	// informed allocation decision when possible.
-	return nil, nil
+	log.Printf("[%s] GetPreferredAllocation called with %d container request(s)", dpi.deviceName, len(in.ContainerRequests))
+
+	response := &pluginapi.PreferredAllocationResponse{}
+
+	for idx, req := range in.ContainerRequests {
+		log.Printf("[%s] Container request %d: Available devices=%v, MustInclude=%v, AllocationSize=%d",
+			dpi.deviceName, idx, req.AvailableDeviceIDs, req.MustIncludeDeviceIDs, req.AllocationSize)
+
+		// Build a map of device ID to NUMA node from our device list
+		deviceToNUMA := make(map[string]int64)
+		for _, dev := range dpi.devs {
+			if dev.Topology != nil && len(dev.Topology.Nodes) > 0 {
+				deviceToNUMA[dev.ID] = dev.Topology.Nodes[0].ID
+			}
+		}
+
+		// Group available devices by NUMA node
+		numaToDevices := make(map[int64][]string)
+		for _, deviceID := range req.AvailableDeviceIDs {
+			numaNode := deviceToNUMA[deviceID]
+			numaToDevices[numaNode] = append(numaToDevices[numaNode], deviceID)
+		}
+
+		// Prefer devices from the same NUMA node
+		var preferredDevices []string
+
+		// First, try to satisfy the request from a single NUMA node
+		for numaNode, devices := range numaToDevices {
+			if int32(len(devices)) >= req.AllocationSize {
+				log.Printf("[%s] Found %d devices on NUMA node %d (requested %d)",
+					dpi.deviceName, len(devices), numaNode, req.AllocationSize)
+				// Take the first AllocationSize devices from this NUMA node
+				for i := int32(0); i < req.AllocationSize && i < int32(len(devices)); i++ {
+					preferredDevices = append(preferredDevices, devices[i])
+				}
+				break
+			}
+		}
+
+		// If we couldn't find enough devices on a single NUMA node, use available devices
+		if len(preferredDevices) == 0 {
+			log.Printf("[%s] Could not satisfy request from single NUMA node, using available devices",
+				dpi.deviceName)
+			for i := int32(0); i < req.AllocationSize && i < int32(len(req.AvailableDeviceIDs)); i++ {
+				preferredDevices = append(preferredDevices, req.AvailableDeviceIDs[i])
+			}
+		}
+
+		// Add MustIncludeDeviceIDs if specified
+		mustIncludeMap := make(map[string]bool)
```

Comment:
maybe we need prefer "must include" devices before filling preferred list with available ones? And only after all "must include" devices added to preferred list then append available devices until we meet allocation size? Current code forms preferred list with available devices with the total number of devices equals to AllocationSize and then appends must-have devices atop of this list, making its size > AllocationSize.

Reply:
Good point! Right now we build preferredDevices to exactly AllocationSize from the available pool and only afterwards tack on anything from MustIncludeDeviceIDs, so the list can grow beyond the requested size, this breaks the contract Topology Manager expects: the response should contain exactly AllocationSize entries, with the must-include devices guaranteed to be part of that set.
so I refined `GetPreferredAllocation` so must-include IDs are deduped, inserted first, and the response never exceeds AllocationSize. The requests with more must-includes than slots now fail fast. Node selection now prioritizes NUMA nodes already represented by must-includes, preserves the kubelet’s device order when a single node cannot satisfy the request, and logs how remaining slots get filled.